app/fileManager.js

```diff
@@ -60,16 +73,63 @@ export default function(state, emitter) {
     emitter.emit('render');
   }
 
+  /*
+
+  function updateDloads() {
+    emitter.emit('updateDloads');
+  }
+*/
+
+  /*const dlimitArray = [];
+  const fileIdArray = [];
+  function ifRerender(file) {
+    if (fileIdArray.indexOf(file.id) === -1) {
+      const fileDetails = {
+        fileId: file.id,
+        dlimit: file.dlimit,
+        dtotal: file.dtotal
+      };
+      fileIdArray.push(file.id);
+      dlimitArray.push(fileDetails);
+      return false;
+    } else {
+      for (const fileDetail of dlimitArray) {
+        if (file.id === fileDetail.fileId) {
+          if (
+            file.dlimit !== fileDetail.dlimit ||
+            file.dtotal !== fileDetail.dtotal
+          ) {
+            fileDetail.dtotal = file.dtotal;
+            fileDetail.dlimit = file.dlimit;
+            return true;
+          } else {
+            return false;
+          }
+        }
+      }
+    }
+  }*/
+
   async function checkFiles() {
     const files = state.storage.files;
+
     let rerender = false;
+
     for (const file of files) {
-      const ok = await exists(file.id);
-      if (!ok) {
+      const receivedFile = await getDLCounts(file);
+      if (!receivedFile) {
         state.storage.remove(file.id);
         rerender = true;
       }
+
+      if (
+        file.dlimit !== receivedFile.dlimit ||
```

Comment:
Might `file` and `receivedFile` both be references to the same object?

Reply:
`file` has been picked from state.storage.files whereas receivedFile is the one we got from `FileReceiver.js`.

Logging them on the console brings this 

![sae](https://user-images.githubusercontent.com/25258877/35134637-9e02c1be-fcfd-11e7-80b8-41365bce91fd.png)


This shows them as exact identical. Even after doing a few downloads. What else can be the strategy to check for the same?

Reply:
So we know that `file` and `receivedFile` are the same object here. We also know that `getDLCounts` gets the latest counts. We want to compare the latest counts to the previous counts. When (as in line in this function) do we know the previous counts, and can we remember them so we can compare them here?

It might help to step through it in the debugger

Reply:
Are you hinting at `state.storage.getFileById(file.id)` ?

Reply:
> Are you hinting at state.storage.getFileById(file.id) ?

nope :)

think about this sample code:
```js
function foo(obj) {
  obj.a = obj.a + 1;
  return obj;
}

const x = { a: 1 };
console.log(x.a);

const y = foo(x);

console.log(y.a, x.a);

// we want to compare the _old_ value of x.a with y.a
if (x.a !== y.a) {
  console.log('this will never happen because x === y');
}
// when do we know the _old_ value of x.a? (when it was 1)
```

Reply:
Apologies for the delay.

In this code, according to me, the old value of x.a is known at the 6th line of the above code OR basically, just at the function call. An equivalent of this in my code should right before when I send a file to send . So that means the previous commit of comparing file and receiverFile should work. Though we have seen that it doesnt work. Would you suggest something different??

Reply:
```js
  async function checkFiles() {
    const files = state.storage.files;

    let rerender = false;

    for (const file of files) {
      const oldLimit = file.dlimit;
      const oldTotal = file.dtotal;
      const receivedFile = await getDLCounts(file);
      if (!receivedFile) {
        state.storage.remove(file.id);
        rerender = true;
      }

      if (
        oldLimit !== receivedFile.dlimit ||
        oldTotal !== receivedFile.dtotal
      ) {
        rerender = true;
      }
    }

    if (rerender) {
      render();
    }
}
```

Reply:
Thanks a lot @dannycoates. I just never realised that the contents of `file` too are changed after it is passed. I'll implement the same. are there any other changes needed?

Reply:
@dannycoates I've made these changes. Are there any other changes to be made ?